sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/BufferingStreamObserver.java

```diff
@@ -73,94 +83,58 @@ private void drainQueue() {
           if (value != POISON_PILL) {
             outboundObserver.onNext(value);
           } else {
+            outboundObserver.onCompleted();
             return;
           }
         }
         phaser.awaitAdvance(currentPhase);
       }
-    } catch (InterruptedException e) {
-      Thread.currentThread().interrupt();
-      throw new IllegalStateException(e);
+    } catch (OnErrorException e) {
+      outboundObserver.onError(e.getCause());
```

Comment:
As far as I can tell, this spot in the code is the only reason to have `OnErrorException`, and the only difference seems to be to avoid a redundant `queue.cancel` (it gets cancelled in `onError` and then again here). Is calling `queue.cancel` twice enough of a problem to warrant this?

Reply:
Ack. I was just thinking that the general Exception catch already calls `outboundObserver.onError(e)`, so if you cancel  the queue with the top-level exception it would be the same. Or am I misunderstanding what happens with `getCause`? I'm reading it as getting the original exception that was nested in the `OnErrorException`, but maybe it's a more top level exception than that?

Regardless, this is all nit-picking at this point. Not worth blocking approval on.